{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "**Recreating the plasmid figure**\n",
        "\n",
        "\n",
        "The plasmid figure was done by running a multifasta file containing all plasmids in a database, and then recording how many of those plasmids were found by that method.\n",
        "\n",
        "1. First, run your multifasta file of plasmids in skandiver, genomad, and mefinder. In the paper, the following database was used: https://datadryad.org/stash/dataset/doi:10.15146/R33X2J\n",
        "\n",
        "The following commands were used to run skandiver, mefinder, and genomad on a multifasta file or directory containing all plasmids from that database:\n",
        "\n",
        "\n",
        "```\n",
        "#skandiver\n",
        "./skandiver.sh all_plasmids/ all_plasmid_10k_results 10000 gtdb_skani_database_ani\n",
        "\n",
        "#mefinder\n",
        "mefinder find --contig all_plasmids/all_plasmids.fasta mefinder_all_plasmids_output\n",
        "\n",
        "#genomad\n",
        "genomad end-to-end --cleanup --splits 8 all_plasmids/all_plasmids_with_names.fasta genomad_all_plasmids_results genomad_db\n",
        "```\n",
        "2. Next, we will extract the names of all plasmids that were found by each method. We use the following command:\n",
        "\n",
        "```\n",
        "python3 /usr1/shared/find_gene.py plasmid_names.txt output_file.txt/.csv found_plasmids.txt\n",
        "```\n",
        "The first argument is a .txt file of the names of all plasmids, each on a new line. The second argument is the output file of the method you used, and the third is the name you want of your output file.\n",
        "\n",
        "3. Now that all data has been collected from each method, we can graph the results. You will need a .csv file with the plasmid name and its length. Use the following code block / function below to plot the data:\n"
      ],
      "metadata": {
        "id": "y8ESm5YuDT11"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# if you don't have the data for one of the methods, just say None in its place\n",
        "plot_hist(<plasmid_lengths>, <skandiver>, <genomad>, <mefinder>, <bucket_num(number of buckets you want in the histogram>)"
      ],
      "metadata": {
        "id": "ot5kUXZ1HqhD"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Graph Plotting Function\n",
        "import math\n",
        "import matplotlib.pyplot as plt\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "import matplotlib.patches as patches\n",
        "\n",
        "def make_percentages(og_bucket_count, new_bucket):\n",
        "    percentages = []\n",
        "    for i in range(len(og_bucket_count)):\n",
        "        per = (new_bucket[i]/og_bucket_count[i])*100\n",
        "        #print(\"per\", per)\n",
        "        percentages.append(per)\n",
        "    return percentages\n",
        "\n",
        "def obtain_bucket_counts(plasmid_lengths_df, bin_edges, found_plasmids, num_buckets):\n",
        "    found_bucket_counts = {i: 0 for i in range(num_buckets)}\n",
        "    for length in plasmid_lengths_df[plasmid_lengths_df[\"Gene Name\"].isin(found_plasmids)][\"Length\"]:\n",
        "        bucket_index = find_index(length, bin_edges)\n",
        "        #print(length)\n",
        "        found_bucket_counts[bucket_index] += 1\n",
        "    return found_bucket_counts\n",
        "\n",
        "def find_index(length, bin_edges):\n",
        "    for i in range(0, len(bin_edges)-1):\n",
        "        if bin_edges[i] <= length and length <= bin_edges[i+1]:\n",
        "            #print(f'{bin_edges[i]} <={length} <= {bin_edges[i+1]}', i)\n",
        "            return i\n",
        "    return len(bin_edges)-2\n",
        "\n",
        "def plot_hist(plasmid_lengths, skandiver, genomad, mefinder, bucket_num):\n",
        "    # Sample data for four different sets of bars\n",
        "    skani_color =  (254/255, 17/255, 17/255)\n",
        "    genomad_color =  (1/255, 207/255, 13/255)\n",
        "    mefinder_color =  (1/255, 40/255, 218/255)\n",
        "\n",
        "    bucket_arr = []\n",
        "    for i in range(1, bucket_num+1):\n",
        "        bucket_arr.append(i)\n",
        "    x = np.array(bucket_arr)\n",
        "\n",
        "    plasmid_lengths_df = pd.read_csv(plasmid_lengths)\n",
        "\n",
        "    min_length = min(plasmid_lengths_df[\"Length\"])\n",
        "    max_length = max(plasmid_lengths_df[\"Length\"])\n",
        "    # Specify the number of buckets and bucket size\n",
        "    num_buckets = bucket_num\n",
        "    bin_edges = np.logspace(np.log10(min_length), np.log10(max_length), num=num_buckets+1)\n",
        "    bucket_size = max(plasmid_lengths_df[\"Length\"]) // num_buckets\n",
        "\n",
        "    bucket_counts_original = {i: 0 for i in range(num_buckets)}\n",
        "    for length in plasmid_lengths_df[\"Length\"]:\n",
        "        bucket_index = find_index(length, bin_edges)\n",
        "        bucket_counts_original[bucket_index] += 1\n",
        "\n",
        "    # Width of each bar\n",
        "    width = 0.2\n",
        "\n",
        "    genomad_color =  (1/255, 207/255, 13/255)\n",
        "\n",
        "    if skandiver:\n",
        "        with open(skandiver, \"r\") as file:  # Replace with the path to your found plasmids file\n",
        "            found_skani_plasmids = [line.strip() for line in file]\n",
        "        skandiver_count = obtain_bucket_counts(plasmid_lengths_df, bin_edges, found_skani_plasmids, bucket_num)\n",
        "        skandiver_plot = make_percentages(bucket_counts_original, skandiver_count)\n",
        "        y3 = np.array(skandiver_plot)\n",
        "        plt.bar(x + 3 * width, y3, width=width, color='red', label='')\n",
        "\n",
        "    if mefinder:\n",
        "            with open(mefinder, \"r\") as file:  # Replace with the path to your found plasmids file\n",
        "                found_mefinder_plasmids = [line.strip() for line in file]\n",
        "            mefinder_count = obtain_bucket_counts(plasmid_lengths_df, bin_edges, found_mefinder_plasmids, bucket_num)\n",
        "            mefinder_plot = make_percentages(bucket_counts_original, mefinder_count)\n",
        "            y2 = np.array(mefinder_plot)\n",
        "            # Plotting the second set of bars with a specific color\n",
        "            plt.bar(x + width, y2, width=width, color='blue')\n",
        "\n",
        "    if genomad:\n",
        "            with open(genomad, \"r\") as file:  # Replace with the path to your found plasmids file\n",
        "                found_genomad_plasmids = [line.strip() for line in file]\n",
        "            genomad_count = obtain_bucket_counts(plasmid_lengths_df, bin_edges, found_genomad_plasmids, bucket_num)\n",
        "            genomad_plot = make_percentages(bucket_counts_original, genomad_count)\n",
        "            y4 = np.array(genomad_plot)\n",
        "            # Plotting the fourth set of bars with a specific color\n",
        "            plt.bar(x + 2*width, y4, width=width, color=(1/255, 207/255, 13/255))\n",
        "\n",
        "    #print(\"mefinder\", mefinder_plot)\n",
        "    #print(\"genomad\", genomad_plot)\n",
        "    #print(\"skandiver\", skandiver_plot)\n",
        "    bucket_labels = []\n",
        "    for i in range(bucket_num):\n",
        "        bucket_labels.append(f\"{bin_edges[i]}-{bin_edges[i+1]}\")\n",
        "    print(\"bin edges \", bucket_labels)\n",
        "    # Labeling the x-axis\n",
        "    plt.xlabel('Bin edges')\n",
        "\n",
        "    # Labeling the y-axis\n",
        "    plt.ylabel('Percent Plasmids Found')\n",
        "\n",
        "    # Adding a title\n",
        "    plt.title('Percent of Plasmids Found For each Method')\n",
        "\n",
        "    plt.legend(handles=[patches.Patch(color=skani_color, label='skandiver'),\n",
        "                    patches.Patch(color=mefinder_color, label='mefinder'),\n",
        "                    patches.Patch(color=genomad_color, label='genomad')],\n",
        "            bbox_to_anchor=(1.05, 1), loc='upper left')\n",
        "\n",
        "\n",
        "    # Display the plot\n",
        "    plt.show()"
      ],
      "metadata": {
        "id": "LBSJ261OHlMm"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}